### PR TITLE
Cache the current head to eliminate eth_getBlockNumber calls in ethTx

### PR DESCRIPTION
**Pull request by j16r**
The latest head has already been passed to us via the OnNewHead callback, or can be obtained from the Heads table, use these instead of making another on chain call.

### CODE REVIEW

File: core/store/eth_client.go [hunk 1]
```diff
@@ -12,6 +12,23 @@ import (
 	"github.com/smartcontractkit/chainlink/core/utils"
 )
 
+//go:generate mockgen -package=mocks -destination=../internal/mocks/eth_wrapper_mocks.go github.com/smartcontractkit/chainlink/core/store EthWrapper
+
+// EthWrapper is the interface supplied the EthClient
```
Comment (felder-cl) [inline core/store/eth_client.go]: Looks like there might be a typo here.  "EthWrapper is the interface supplied by(?) the EthClient"?

File: core/store/eth_client.go [hunk 2]
```diff
@@ -12,6 +12,23 @@ import (
...
+	GetEthBalance(address common.Address) (*assets.Eth, error)
+	GetERC20Balance(address common.Address, contractAddress common.Address) (*big.Int, error)
+	SendRawTx(hex string) (common.Hash, error)
+	GetTxReceipt(hash common.Hash) (*models.TxReceipt, error)
+	GetBlockByNumber(hex string) (models.BlockHeader, error)
+	GetLogs(q ethereum.FilterQuery) ([]models.Log, error)
+	GetChainID() (*big.Int, error)
+	SubscribeToLogs(channel chan<- models.Log, q ethereum.FilterQuery) (models.EthSubscription, error)
+	SubscribeToNewHeads(channel chan<- models.BlockHeader) (models.EthSubscription, error)
+}
```
Comment (felder-cl) [inline core/store/eth_client.go]: I feel a more descriptive name could be used, though I'm struggling to come up with a suggestion.  Maybe something like 'EthProvider' or 'EthServicer' or 'EthQuerier'?

File: core/store/tx_manager_test.go [hunk 3]
```diff
@@ -22,30 +24,35 @@ import (
...
 	t.Parallel()
-	app, cleanup := cltest.NewApplicationWithKey(t)
+
+	store, cleanup := cltest.NewStore(t)
 	defer cleanup()
-	store := app.Store
-	manager := store.TxManager
 
-	from := cltest.GetAccountAddress(t, app.GetStore())
+	ctrl := gomock.NewController(t)
```
Comment (se3000) [inline core/store/tx_manager_test.go]: Do we ever need `ctrl` outside of NewMockEthClient? Wondering if it'd be worth saving a line of boilerplate with a test helper.

Comment (j16r) [reply core/store/tx_manager_test.go]: You need to call `ctrl.Finish` wherever you use this and it's now used in a bunch of tests other than the MockEthCallerSubscriber so it wouldn't make much of a difference.

File: core/internal/cltest/cltest.go [hunk 4]
```diff
@@ -86,10 +86,8 @@ func NewConfig(t testing.TB) (*TestConfig, func()) {
 	return NewConfigWithWSServer(t, wsserver), cleanup
 }
 
-// NewConfigWithWSServer return new config with specified wsserver
-func NewConfigWithWSServer(t testing.TB, wsserver *httptest.Server) *TestConfig {
-	t.Helper()
-
+// NewTestConfig returns a test configuration
+func NewTestConfig(t testing.TB) *TestConfig {
```
[the remark this answers is not in the document]
Comment (j16r) [reply core/internal/cltest/cltest.go]: Added in.

**Comment by j16r** [comment]
The one drawback I've noticed of the new mocks package (it's not that new, we have it in around 10 tests already) is that you can't assert all calls have been made more than once. It only has a `Finish` method which is meant to be called once at the end of the test. This might be a simple PR.

**Pull request merged**